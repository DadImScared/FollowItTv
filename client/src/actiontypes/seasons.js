
export const ADD_SEASONS = 'seasons/ADD_SEASONS';
export const ADD_SEASON_EPISODES = 'seasons/ADD_SEASON_EPISODES';
