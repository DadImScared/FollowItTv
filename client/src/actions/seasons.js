
import axios from 'axios';
import { ADD_SEASONS, ADD_SEASON_EPISODES } from '../actiontypes/seasons';
import { loadingStart, loadingEnd } from './loading';
import { seasonEpisodesLoading } from '../loadingIds';


export const baseUrl = 'http://api.tvmaze.com/';

export const addSeasons = (showId, seasons) => {
  return {
    type: ADD_SEASONS,
    showId,
    seasons
  };
};

export const addSeasonEpisodes = (seasonId, episodes) => {
  return {
    type: ADD_SEASON_EPISODES,
    seasonId,
    episodes
  };
};

export const getSeasons = async (showId) => {
  return await axios.get(`${baseUrl}shows/${showId}/seasons`);
};

export const getSeasonEpisodes = async (seasonId) => {
  return await axios.get(`${baseUrl}seasons/${seasonId}/episodes`);
};

// work in progress
export const findSeasonEpisodes = (seasonId, showId) => {
  return async (dispatch) => {
    const loadingId = seasonEpisodesLoading(seasonId);
    dispatch(loadingStart(loadingId));
    const { data } = await axios.get(`${baseUrl}seasons/${seasonId}/episodes`);
  };
};

export const normalizeSeasonEpisodes = (showId, episodes) => {
  const episodeList = [];
  const newEpisodes = {};
  episodes.forEach((episode) => {
    episodeList.push(episode.id);
    newEpisodes[episode.id] = {
      ...episode,
      show: showId
    };
  });

  return {
    episodeList,
    newEpisodes
  };
};
