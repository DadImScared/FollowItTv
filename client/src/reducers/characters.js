
import { ADD_CHARACTERS } from '../actiontypes/characters';

function characters(state = {}, action) {
  switch(action.type) {
    case ADD_CHARACTERS:
      return {
        ...state,
        ...action.characters
      };
    default:
      return state;
  }
}

export default characters;
