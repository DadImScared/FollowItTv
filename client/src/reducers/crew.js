
import { ADD_CREW } from '../actiontypes/crew';

function crew(state = {}, action) {
  switch (action.type) {
    case ADD_CREW:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          ...action.crewMembers
        }
      };
    default:
      return state;
  }
}

export default crew;
