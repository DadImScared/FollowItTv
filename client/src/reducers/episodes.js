
import { ADD_EPISODES } from '../actiontypes/episodes';


function episodes(state={}, action) {
  switch (action.type) {
  case ADD_EPISODES:
    return {
      ...state,
      ...action.episodes
    };
  default:
    return state;
  }
}

export const getEpisodes = (state) => state.episodes;

export default episodes;
