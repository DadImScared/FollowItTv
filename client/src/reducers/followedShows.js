
import * as followedShowsTypes from '../actiontypes/followedShows';
import { followShow, unfollowShow } from './followedShowsUtility';

// const updateShow = (showsState, action) => {
//   return showsState[action.id] ? {
//     ...showsState,
//     [action.id]: {
//       ...showsState[action.id],
//       isFollowing: action.isFollowing
//     }
//   }:{
//     ...showsState,
//     [action.id]: {
//       ...action.payload,
//       isFollowing: action.isFollowing
//     }
//   };
// };

// const followShow = (showState, action) => {
//   return showState.concat(action.id);
// };
//
// const unfollowShow = (showState, action) => {
//   return showState.filter((item) => item !== action.id);
// };

const updateShowInDay = (dayState, action, updateCb)=> {
  const updatedDays = action.days.map((day) => {
    if (dayState[day]) {
      return {
        [day]: updateCb(dayState[day], action)
      };
    }
  }).reduce((obj, item) => ({ ...obj, ...item }), {});
  return {
    ...dayState,
    ...updatedDays
  };
};

function followedShows(state={}, action) {
  switch(action.type) {
  case followedShowsTypes.ADD_DAY:
    return {
      ...state,
      [action.day]: action.shows
    };
  case followedShowsTypes.UNFOLLOW_SHOW:
    return updateShowInDay(state, action, unfollowShow);
  case followedShowsTypes.FOLLOW_SHOW:
    return updateShowInDay(state, action, followShow);
  default:
    return state;
  }
}

export default followedShows;
