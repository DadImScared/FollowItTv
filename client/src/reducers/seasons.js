
import { createSelector } from 'reselect';

import { ADD_SEASONS, ADD_SEASON_EPISODES } from '../actiontypes/seasons';
import { getEpisodes } from './episodes';
import { getLoading } from './loading';

const initialState = {
  byId: {},
  seasonEpisodes: {}
};

function seasons(state=initialState, action) {
  switch (action.type) {
    case ADD_SEASONS:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.showId]: action.seasons
        }
      };
    case ADD_SEASON_EPISODES:
      return {
        ...state,
        seasonEpisodes: {
          ...state.seasonEpisodes,
          [action.seasonId]: action.episodes
        }
      };
    default:
      return state;
  }
}

export const getSeasonEpisodes = (state, { season: { id } }) => state.seasons.seasonEpisodes[id] || [];

export const makeSeasonEpisodes = () => {
  return createSelector(
    [getSeasonEpisodes, getEpisodes, getLoading('SHOW_SEASONS_', 'season.id')],
    (seasonEpisodes, episodes, isLoading) => ({
      episodes: seasonEpisodes.map((episodeId) => episodes[episodeId]),
      isLoading,
      hasData: seasonEpisodes.length
    })
  );
};

export default seasons;
