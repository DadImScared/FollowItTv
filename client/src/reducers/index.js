
import { combineReducers } from 'redux';

import followedShows from './followedShows';
import followedShowsById from './followedShowsId';
import shows from './shows';
import schedule from './schedule';
import users from './users';
import episodes from './episodes';
import seasons from './seasons';
import search from './search';

const reducer = combineReducers({
  followedShows,
  followedShowsById,
  seasons,
  episodes,
  shows,
  schedule,
  users,
  search
});

export default reducer;
