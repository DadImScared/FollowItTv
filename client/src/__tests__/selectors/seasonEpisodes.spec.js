
import { makeSeasonEpisodes } from '../../reducers/seasons';

describe('seasonEpisodes', () => {
  test('makeSeasonEpisodes', () => {
    const state = {
      seasons: {
        seasonEpisodes: {
          1: [1]
        }
      },
      episodes: {
        1: {
          id: 1,
          name: 'episode 1'
        }
      },
      loading: {
        'SHOW_SEASONS_1': false
      }
    };
    const props = {
      season: { id: 1 }
    };

    const getSeasonEpisodes = makeSeasonEpisodes();
    expect(getSeasonEpisodes.recomputations()).toEqual(0);
    const results = getSeasonEpisodes(state, props);
    getSeasonEpisodes(state, props);
    expect(getSeasonEpisodes.recomputations()).toEqual(1);
    expect(results).toMatchSnapshot();
  });
});
