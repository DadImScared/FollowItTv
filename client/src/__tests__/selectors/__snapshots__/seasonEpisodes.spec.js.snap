// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seasonEpisodes makeSeasonEpisodes 1`] = `
Object {
  "episodes": Array [
    Object {
      "id": 1,
      "name": "episode 1",
    },
  ],
  "hasData": 1,
  "isLoading": false,
}
`;
